File: 1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.js
/**
 * // Definition for a _Node.
 * function _Node(val) {
 *    this.val = val;
 *    this.left = null;
 *    this.right = null;
 *    this.parent = null;
 * };
 */

/**
 * @param {_Node} p
 * @param {_Node} q
 * @return {_Node}
 */
var lowestCommonAncestor = function(p, q) {
    
    function findRoot(node) {
        while(node.parent !== null) {
            node = node.parent
        }
        return node
    }

    const root = findRoot(p)

    function recurse(node, p, q) {
        if (node === null) return null
        if (node === p || node === q) return node

        const left = recurse(node.left, p, q)
        const right = recurse(node.right, p, q)

        if (left && right) return node

        return left || right
    }

    return recurse(root, p, q)
};

File: 1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        p_ptr, q_ptr = p, q

        while p_ptr != q_ptr:
            p_ptr = p_ptr.parent
            if not p_ptr:
                p_ptr = q

            q_ptr = q_ptr.parent
            if not q_ptr:
                q_ptr = p

        return p_ptr